### ISSUE TITLE
upstream `apex.optimizers.FusedAdam` to replace `torch.optim.AdamW`

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#71274]
### 🚀 The feature, motivation and pitch

After running several benchmarks [1](https://github.com/huggingface/transformers/issues/14608#issuecomment-1005219385) and [2](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005220263) it appears that  `apex.optimizers.FusedAdam` is 10-15% faster than `torch.optim.AdamW` (in an ensemble of the HF Trainer loop). 

I'm proposing to replace `torch.optim.AdamW` with the faster `apex.optimizers.FusedAdam` implementation and not require a user to manually build apex, as the latter is not always simple or even possible, and gain a noticeable training speed up.

or perhaps having `torch.optim.FusedAdamW` as an addition and not a replacement if there is a reason to keep the original unfused implementation.

Thank you!

cc @vincentqb @jbschlosser @albanD

**Comment by ngimel** [comment]
There definitely is a need to keep original non-fused implementation, apex FusedAdam doesn't cover all the functionality that regular optimizers provide (e.g. group handling or situations where some params are not getting gradients, or sparse gradients). 
Can you please check if `_multi_tensor` implementation provides speed-ups, as @albanD suggests? It is expected to be slower than apex's fully fused one, but it might be good enough.

**Comment by stas00** [comment]
I did benchmark the `_multi_tensor` feature when it came out a year ago but saw no difference: https://github.com/huggingface/transformers/issues/9965

Let me re-test it again. And I will follow up if it's any better.

**Comment by ngimel** [comment]
That's weird, typically we see good e2e improvements from it. Would you be able to collect a profile from it?

**Comment by stas00** [comment]
I see now that I was benchmarking [HF's Adam vs `torch.optim._multi_tensor.AdamW`](https://github.com/huggingface/transformers/issues/9965#issuecomment-772009817) so need to first re-run checks on`torch.optim.AdamW`  vs `torch.optim._multi_tensor.AdamW`.

HF's Adam is slightly faster than torch's https://github.com/huggingface/transformers/issues/15026#issuecomment-1005220263 (but it also is a bit buggy and we just deprecated it). Hence my looking into the best replacement.

I will follow up once I get a chance to re-test.

**Comment by stas00** [comment]
OK, so first 2 up-to-date benchmarks with 2 different models.

## gpt2-medium

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                               14.28 |           1 |            3.02 |
| --optim adamw_torch       |                               14.08 |           0 |            3.02 |
| --optim adamw_apex_fused  |                               15.67 |          11 |            3.02 |

```
Base command: python examples/pytorch/language-modeling/run_clm.py --model_name_or_path gpt2-medium --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --logging_strategy no --save_strategy no --do_train --max_train_samples 1200 --per_device_train_batch_size 4 --num_train_epochs 1 --warmup_steps 8 --block_size 512 --report_to none --fp16 --output_dir output_benchmark --overwrite_output_dir
```

## T5-base

| Variation                 |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:--------------------------|------------------------------------:|------------:|----------------:|
| --optim adamw_torch_fused |                              210.25 |           0 |            2.40 |
| --optim adamw_torch       |                              210.84 |           0 |            2.40 |
| --optim adamw_apex_fused  |                              235.71 |          12 |            2.40 |

```
Base command: python examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 --max_source_length 512 --max_target_length 512 --num_train_epochs 1 --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix translate English to Romanian:  --warmup_steps 50 --max_train_samples 2500 --dataloader_num_workers 2 --output_dir output_benchmark --overwrite_output_dir
```

where:

- `adamw_torch_fused` : `torch.optim._multi_tensor.AdamW`
- adamw_torch`: `torch.optim.AdamW`

As you can see there is barely any difference between the 2

Let's see if I can figure out how to profile it.

```
*** Setup:
Datetime    : 2022-01-18 19:13:07

Software:
transformers: 4.16.0.dev0
torch       : 1.10.1
cuda        : 11.3
python      : 3.8.11

Hardware:
1 GPUs      : NVIDIA GeForce RTX 3090, 23.70GB
```

**Comment by ngimel** [comment]
https://github.com/pytorch/pytorch/blob/9515213070738d9f273b011038e9a248af143522/torch/optim/_multi_tensor/adamw.py#L83 is for some reason using a non-fused operation, but I'm surprised it has such an effect.

**Comment by stas00** [comment]
Has anybody on the pytorch team benchmarked `torch.optim._multi_tensor.AdamW` vs `torch.optim.AdamW`?

**Comment by crcrpar** [comment]
I took profiles of T5 with `torch.optim.AdamW`, `torch.optim._multi_tensor.AdamW`, and `apex.optimizers.FusedAdam(W)` on 3090 after setting `non_blocking=True` in https://github.com/crcrpar/transformers/pull/1/files#diff-ed55888e6665791fe92cc8fc0c499da54f4ace6738551cd9a2591881cda076deR1925 and https://github.com/crcrpar/transformers/pull/1/files#diff-ebaff1224cad711fd3cefb771ce17d1392ae2dfc7f74dc7da9dd014d7642a344R925 that seemed to remove weirdly long `cudaStreamSynchronize` in `torch.optim._multi_tensor.AdamW` and `apex.optimizers.FusedAdam(W)`.

Here's the profile of `torch.optim.AdamW`.
![image](https://user-images.githubusercontent.com/16191443/153674119-1e677713-4f67-44bf-88bf-17c66325e820.png)
You can see that the utilization of `torch.optim.AdamW.step()` is nice (Each vertical blue bar in the row of "97.1 % Kernels" of "Processes (2) -> [820] Python -> CUDA HW..." has almost the same height of that row).
Also, `.step()` occupies ~25ms/~170ms = ~15% of one iteration (meaning forward, backward, and param update).
So fusing across tensors which what `torch.optim._multi_tensor.AdamW` does will not be as beneficial as fusing across both operations and tensors which `apex.optimizers.FusedAdam(W)` does.

**Comment by stas00** [comment]
@crcrpar, I was curious to try your experiment - are you measuring a speed up with `torch.optim.AdamW` when adding `non_blocking=True` as you suggested?

I tried this and getting same or slightly slower results - with t5-small or t5-base on 3090.

```
diff --git a/src/transformers/models/t5/modeling_t5.py b/src/transformers/models/t5/modeling_t5.py
index 3af2a53de..dc09b9baa 100644
--- a/src/transformers/models/t5/modeling_t5.py
+++ b/src/transformers/models/t5/modeling_t5.py
@@ -920,7 +920,7 @@ class T5Stack(T5PreTrainedModel):
             assert self.is_decoder, f"`use_cache` can only be set to `True` if {self} is used as a decoder"

         if attention_mask is None:
-            attention_mask = torch.ones(batch_size, mask_seq_length).to(inputs_embeds.device)
+            attention_mask = torch.ones(batch_size, mask_seq_length).to(inputs_embeds.device, non_blocking=True)
         if self.is_decoder and encoder_attention_mask is None and encoder_hidden_states is not None:
             encoder_seq_length = encoder_hidden_states.shape[1]
             encoder_attention_mask = torch.ones(
diff --git a/src/transformers/trainer.py b/src/transformers/trainer.py
index 08c83f9f7..16c60ab54 100755
--- a/src/transformers/trainer.py
+++ b/src/transformers/trainer.py
@@ -1917,7 +1917,7 @@ class Trainer:
         elif isinstance(data, (tuple, list)):
             return type(data)(self._prepare_input(v) for v in data)
         elif isinstance(data, torch.Tensor):
-            kwargs = dict(device=self.args.device)
+            kwargs = dict(device=self.args.device, non_blocking=True)
             if self.deepspeed and data.dtype != torch.int64:
                 # NLP models inputs are int64 and those get adjusted to the right dtype of the
                 # embedding. Other models such as wav2vec2's inputs are already float and thus
```

**Comment by crcrpar** [comment]
@stas00 I didn't see slow down in `train_steps_per_second`. I collected some numbers in https://gist.github.com/crcrpar/df7e8537f003e813ffe51d104c04dfd3 by running `python examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 --max_source_length 512 --max_target_length 512 --num_train_epochs 1 --source_lang en --target_lang ro --dataset_name wmt16 --dataset_config ro-en --source_prefix "translate English to Romanian:" --warmup_steps 50 --max_train_samples 2500 --dataloader_num_workers 2 --output_dir /tmp/translation --overwrite_output_dir --optim adamw_torch`.

**Comment by stas00** [comment]
apologies I didn't follow up right away, @crcrpar 

I thought it would make things faster.

So the current status is: `torch.optim._multi_tensor.AdamW` is not faster than `torch.optim.AdamW` even though it's supposed to be faster.

and so the purpose of this Issue remains - `apex.optimizers.FusedAdam` is faster than the 2 former ones.

Please correct me if my conclusion is wrong. Thank you!

**Comment by janeyx99** [comment]
This should be completed with AdamW(...fused=True) now.
